internals/api/auth.go

```diff
@@ -0,0 +1,249 @@
+package api
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/secrethub/secrethub-go/internals/api/uuid"
+)
+
+// AuthMethod options
+const (
+	AuthMethodAWSSTS = "aws-sts"
+)
+
+// SessionType options
+const (
+	SessionTypeHMAC SessionType = "hmac"
+)
+
+// Errors
+var (
+	ErrInvalidSessionType  = errAPI.Code("invalid_session_type").StatusError("invalid session type provided for authentication request", http.StatusBadRequest)
+	ErrInvalidPayload      = errAPI.Code("invalid_payload").StatusError("invalid payload provided for authentication request", http.StatusBadRequest)
+	ErrInvalidAuthMethod   = errAPI.Code("invalid_auth_method").StatusError("invalid auth method", http.StatusBadRequest)
+	ErrMissingField        = errAPI.Code("missing_field").StatusErrorPref("request is missing field %s", http.StatusBadRequest)
+	ErrSessionNotFound     = errAPI.Code("session_not_found").StatusError("session could not be found, it might have expired", http.StatusForbidden)
+	ErrSessionExpired      = errAPI.Code("session_expired").StatusError("session has expired", http.StatusForbidden)
+	ErrAuthFailed          = errAPI.Code("auth_failed").StatusError("authentication failed", http.StatusForbidden)
```

Comment:
This is returned if AWS returns a `http.StatusForbidden`. I haven't found a case where this happens. I first thought this would be if authentiction is incorrect, but then AWS returns a BadRequest. 

So we could opt to drop this.

Reply:
Hmm interesting. So could this happen when an account doesn't have permissions to call AWS STS? If the case never happens, it's redundant to have a specific error for it. If it does, it's good to keep it and make sure we have the correct error code.

Reply:
_Everyone_ has permissions to call `GetCallerIdentity`: 

>No permissions are required to perform this operation. If an administrator adds a policy to your IAM user or role that explicitly denies access to the sts:GetCallerIdentity action, you can still perform this operation. Permissions are not required because the same information is returned when an IAM user or role is denied access. To view an example response, see I Am Not Authorized to Perform: iam:DeleteVirtualMFADevice. 

If you are disallowed to do something, an `AwsAccessDeniedException` is returned.